sequencing-server/src/utils/hasura.ts

```diff
@@ -0,0 +1,314 @@
+  }
+
+  throw new Error('Could not determine the user sending the request');
+}
+
+/**
+ * A helper function to get the username of the user who sent the request.
+ *
+ * @param authSessionVariables If the request has gone through a hasura action this object will be populated.
+ * @param authHeader If the user sends a POST request directly to the endpoint with a bearer token this will be populated.
+ * @returns The username of the user sending the request.
+ */
+export function getUsername(
+  authSessionVariables: AuthSessionVariables | undefined,
+  authHeader: string | undefined,
+): string {
+  if (authSessionVariables !== undefined) {
+    return authSessionVariables['x-hasura-user-id'];
+  } else if (authHeader !== undefined) {
+    return (jwtDecode(authHeader) as Jwt).username;
+  }
+
+  throw new Error('Could not determine the user sending the request');
+}
+
+/**
+ * This function checks to see if request coming from the Hasura action is being executed
+ * by a user with enough permissions. To check this we need to call a GQL query with the
+ * action key and the user session. If we ever merge the databases this should be changed
+ * to call the db function that does the permission checking directly.
+ *
+ * Some endpoints don't need to be checked which is why we have the mappings of endpoints
+ * to actions to check.
+ *
+ * @param url The endpoint that we're checking the permissions for.
+ * @param graphqlClient The GQL Client we're using to make the permission check.
+ * @returns True if the user is able to call the given action, false otherwise.
+ */
+export async function canUserPerformAction(
+  url: string,
+  graphqlClient: GraphQLClient,
+  hasuraSession: AuthSessionVariables,
+  body: any,
+): Promise<boolean> {
+  const role = hasuraSession['x-hasura-role'];
+  const username = hasuraSession['x-hasura-user-id'];
+
+  // The aerie_admin role always has NO_CHECK permissions on all actions.
+  if (role === 'aerie_admin') {
+    return true;
+  }
+
+  const permissionCheckQuery = await graphqlClient.request(
+    gql`
+      query checkUserPermissions($role: user_roles_enum!, $key: String!) {
+        user_role_permission_by_pk(role: $role) {
+          action_permissions(path: $key)
+        }
+      }
+    `,
+    {
+      role,
+      key: ENDPOINTS_TO_ACTION_KEY[url],
+    },
+  );
+
+  const permission = permissionCheckQuery.user_role_permission_by_pk.action_permissions as HasuraPermissions | null;
+
+  if (permission === null) {
+    return false;
+  }
+
+  const simulationDatasetId = body.input.simulationDatasetId as number;
+  let missionModelId = body.input.missionModelId as number;
+  let planId = body.input.planId as number;
+
+  // If we have a missionModelId and we need to check a plan permission, get the planId.
+  if (missionModelId !== null && permission in HASURA_PLAN_PERMISSIONS) {
+    planId = await getPlanId(graphqlClient, missionModelId);
+  }
```

Comment:
This will not work, as one mission model is associated with many plans, and this call to getPlanId only returns one of those planIds.

Reply:
The following queries will properly check if a user is a Plan Owner/Collaborator for any plan belonging to a mission model. They return `null` if the mission model id is invalid. Otherwise, they return an object containing an array that's either empty or has a single element.

```gql
query planOwner($id: Int!, $username: String!) {
  mission_model_by_pk(id: $id) {
    plans(where: {owner: {_eq: $username}}, limit: 1) {
      name
    }
  }
}

query planCollaborator($id: Int!, $username: String!) {
  mission_model_by_pk(id: $id) {
    plans(where: {collaborators: {collaborator: {_eq: $username}}}, limit: 1) {
      name
    }
  }
}
```